### PR TITLE
fix: NPE in EncryptionImpl

### LINKED ISSUE #1139: NPE in EncryptionServiceImpl
### Bug Report    

**Actual Behaviour**

Following error log is generated due to Null Pointer Exception.

**Steps to reproduce it**

Build and run the app in phone.

**LogCat for the issue**

`E/EncryptionServiceImpl: java.lang.NullPointerException: Attempt to invoke virtual method 'byte[] java.lang.String.getBytes()' on a null object reference
                             at android.util.Base64.decode(Base64.java:118)
                             at org.fossasia.openevent.app.data.encryption.EncryptionServiceImpl.decryptString(EncryptionServiceImpl.java:114)
                             at org.fossasia.openevent.app.data.encryption.EncryptionServiceImpl.decrypt(EncryptionServiceImpl.java:73)
                             at org.fossasia.openevent.app.core.auth.login.LoginViewModel.getLogin(LoginViewModel.java:98)
                             at org.fossasia.openevent.app.core.auth.login.LoginFragment.onStart(LoginFragment.java:65)
                             at android.support.v4.app.Fragment.performStart(Fragment.java:2369)
                             at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1460)
                             at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1752)
                             at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1821)
                             at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3251)
                             at android.support.v4.app.FragmentManagerImpl.dispatchStart(FragmentManager.java:3210)
                             at android.support.v4.app.FragmentController.dispatchStart(FragmentController.java:206)
                             at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:605)
                             at android.support.v7.app.AppCompatActivity.onStart(AppCompatActivity.java:177)
                             at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1270)
                             at android.app.Activity.performStart(Activity.java:6689)
                             at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2622)
                             at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2724)
                             at android.app.ActivityThread.-wrap12(ActivityThread.java)
                             at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1473)
                             at android.os.Handler.dispatchMessage(Handler.java:102)
                             at android.os.Looper.loop(Looper.java:154)
                             at android.app.ActivityThread.main(ActivityThread.java:6123)
                             at java.lang.reflect.Method.invoke(Native Method)
                             at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:867)
                             at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:757)
E/EncryptionServiceImpl: java.lang.NullPointerException: Attempt to invoke virtual method 'byte[] java.lang.String.getBytes()' on a null object reference
                             at android.util.Base64.decode(Base64.java:118)
                             at org.fossasia.openevent.app.data.encryption.EncryptionServiceImpl.decryptString(EncryptionServiceImpl.java:114)
                             at org.fossasia.openevent.app.data.encryption.EncryptionServiceImpl.decrypt(EncryptionServiceImpl.java:73)
                             at org.fossasia.openevent.app.core.auth.login.LoginViewModel.getLogin(LoginViewModel.java:99)
                             at org.fossasia.openevent.app.core.auth.login.LoginFragment.onStart(LoginFragment.java:65)
                             at android.support.v4.app.Fragment.performStart(Fragment.java:2369)
                             at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1460)
                             at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1752)
                             at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1821)
                             at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3251)
                             at android.support.v4.app.FragmentManagerImpl.dispatchStart(FragmentManager.java:3210)
                             at android.support.v4.app.FragmentController.dispatchStart(FragmentController.java:206)`
                     

**Android version and Phone Model**

Jelly Bean and Moto G4 Plus
    

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by Masquerade0097**
Fixes #1139 

Checklist:

- [x] I have checked for PMD and check-style issues 
- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream development branch.

Changes: Null check for encryption string.

### CODE REVIEW

File: app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java [hunk 1]
```diff
@@ -102,6 +104,10 @@ private String encryptString(String toEncrypt) {
 
     @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
     private String decryptString(String encrypted) {
+        if (Utils.isEmpty(encrypted)) {
+            return null;
+        }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java]: Is encrypt handling the null values?

**Comment by sridharjajoo** [comment]
@Masquerade0097 What is the API level?

**Comment by iamareebjamal** [comment]
@sridharjajoo Doesn't matter, a null string is being passed into the decrypt function as it will be when loading first time

**Comment by sridharjajoo** [comment]
@iamareebjamal Ok. But the app isn't crashing for API >= 18

**Comment by iamareebjamal** [comment]
Because you have already stored encrypted value in shared preferences

**Comment by Masquerade0097** [comment]
@sridharjajoo App isn't crashing in my device too (API Level 24). This error is shown in logcat.

File: app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java (hunk 1 above)
Comment (Masquerade0097) [reply app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java]: Yes, the function encryptString() has a null check.

**Comment by sridharjajoo** [comment]
@Masquerade0097 It will crash for API<18 because  getBytes() is being called on the encrypted string which is apparently null when the app loads inititally. You can change the decryptString( ) used for API < 18 from decryptString(bytes[] cipherText, SecretKey secretkey) to decryptString(String encryptedString, SecretKey secretKey) and then add a null check in this method that if the encryptedString isn't null then only the getBytes should be called on it.

**Comment by Masquerade0097** [comment]
@sridharjajoo Okay, I will change it.

**Comment by Masquerade0097** [comment]
@iamareebjamal @sridharjajoo I have added null check in methods `encrypt` and `decrypt` such that any null checks won't be required in child methods like `decryptString` and `encryptString`.

File: app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java [hunk 2]
```diff
@@ -228,17 +234,14 @@ private void createKeys() throws NoSuchProviderException,
...
-            } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | IllegalBlockSizeException | BadPaddingException | UnsupportedEncodingException e) {
-                Timber.w(e);
-            }
-            return cipherText;
+        try {
+            cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
+            cipher.init(Cipher.ENCRYPT_MODE, secret);
+            cipherText = cipher.doFinal(message.getBytes("UTF-8"));
+        } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | IllegalBlockSizeException | BadPaddingException | UnsupportedEncodingException e) {
+            Timber.w(e);
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java]: null should be returned in case of exception

**Comment by Masquerade0097** [comment]
@iamareebjamal @sridharjajoo Changes done. Please review.

File: app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java [hunk 3]
```diff
@@ -228,17 +234,15 @@ private void createKeys() throws NoSuchProviderException,
...
+            return null;
         }
-        return null;
+        return cipherText;
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java]: Can be returned from try

File: app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java [hunk 4]
```diff
@@ -228,17 +234,15 @@ private void createKeys() throws NoSuchProviderException,
         Cipher cipher;
         byte[] cipherText = new byte[0];
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java]: Both of these can be made local

File: app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java [hunk 5]
```diff
@@ -252,6 +256,7 @@ private String decryptString(byte[] cipherText, SecretKey secret) {
             decryptString = new String(cipher.doFinal(cipherText), "UTF-8");
         } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | IllegalBlockSizeException | BadPaddingException | UnsupportedEncodingException e) {
             Timber.w(e);
+            return null;
         }
         return decryptString;
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java]: Similarly here

File: app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java [hunk 6]
```diff
@@ -231,34 +231,32 @@ private void createKeys() throws NoSuchProviderException,
     *
     */
     private byte[] encryptString(String message, SecretKey secret) {
-        Cipher cipher;
-        byte[] cipherText = new byte[0];
-
         try {
+            Cipher cipher;
+            byte[] cipherText = new byte[0];
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java]: Declare and assign at same time

File: app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java [hunk 7]
```diff
@@ -231,34 +231,32 @@ private void createKeys() throws NoSuchProviderException,
...
     }
 
     private String decryptString(byte[] cipherText, SecretKey secret) {
         /* Decrypt the message, given derived encContentValues and initialization vector. */
-        Cipher cipher = null;
-        String decryptString = null;
-
         try {
+            Cipher cipher = null;
+            String decryptString = null;
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/encryption/EncryptionServiceImpl.java]: Same here, and no need to delare variable decryptString

**Pull request merged**